File: cli/fastchargeapi_cli/auth_file.py
import dataclasses
import json
from dataclasses import dataclass
from functools import cache
from pathlib import Path
from typing import Literal, Optional, Union

import jwt
import requests
from cryptography.x509 import load_pem_x509_certificate

from . import config


@cache
def get_google_cert() -> dict:
    google_cert = requests.get(
        "https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com"
    ).json()
    return google_cert


def get_auth_file_path(profile: Optional[str]) -> Path:
    if profile:
        return Path.home() / f".fastcharge/auth.{profile}.json"
    return Path.home() / ".fastcharge/auth.json"


def list_auth_files() -> list[Path]:
    return list(Path.home().glob(".fastcharge/auth.*.json")) + [
        Path.home() / ".fastcharge/auth.json"
    ]


def delete_auth_file(profile: Optional[str]) -> None:
    auth_file_path = get_auth_file_path(profile)
    if auth_file_path.exists():
        auth_file_path.unlink()


@dataclass
class AuthFileContent:
    id_token: str
    refresh_token: Union[Literal["firebase"], Literal["fastchargeapi"]]
    user_pk: str
    email: str
    issuer: str


def write_to_auth_file(
    *,
    profile: Optional[str] = None,
    issuer: Optional[str] = "firebase",
    id_token: Optional[str] = None,
    refresh_token: Optional[str] = None,
    user_pk: Optional[str] = None,
    email: Optional[str] = None,
) -> AuthFileContent:
    """Partially update the auth file context, or create a new auth file with
    the specified content.

    Profile: The profile to write to. If None, writes to the default profile.
    """
    auth_file_path = get_auth_file_path(profile)
    auth_file_path.parent.mkdir(exist_ok=True)
    auth_file = _read_auth_file(profile)
    data = dataclasses.asdict(auth_file) if auth_file else {}
    data.update(
        {
            k: v
            for k, v in {
                "id_token": id_token,
                "refresh_token": refresh_token,
                "user_pk": user_pk,
                "email": email,
                "issuer": issuer,
            }.items()
            if v is not None
        }
    )
    auth_file_path.write_text(json.dumps(data))
    auth_file_path.chmod(0o600)
    return AuthFileContent(**data)


def read_or_refresh_auth_file(
    profile: Optional[str],
    force_refresh=False,
) -> Optional[AuthFileContent]:
    """
    Gets the id token from the auth file. If the file doesn't exist, returns
    None.

    If the id token is issued by Firebase, verifies the token. If the id token
    is invalid, refreshes it and returns the new id token. If unable to refresh,
    return None.

    If the id token is not issued by FastchargeAPI, skips the verification.

    The new id token and refresh token are written to the auth file.
    """
    if auth := _read_auth_file(profile):
        if auth.issuer == "firebase":
            if user := verify_or_refresh_id_token(
                auth.id_token, auth.refresh_token, force_refresh
            ):
                if user.refreshed:
                    return write_to_auth_file(
                        id_token=user.id_token,
                        refresh_token=user.refresh_token,
                        email=user.email,
                        user_pk=query_user_pk(user.id_token, user.email),
                    )
                else:
                    return auth
        elif auth.issuer == "fastchargeapi":
            return auth
        else:
            raise ValueError(f"Unknown issuer: {auth.issuer}")
    return None


def _read_auth_file(profile: Optional[str]) -> Optional[AuthFileContent]:
    """Read the decoded id token from the auth file. If the file doesn't exist,
    return None."""
    auth_file = get_auth_file_path(profile)
    if auth_file.exists():
        auth = json.loads(auth_file.read_text())
        auth.setdefault("issuer", "firebase")  # old auth file doesn't have issuer
        return AuthFileContent(**auth)
    else:
        return None


def query_user_pk(id_token: str, email: str) -> str:
    from .__generated__ import gql_operations as GQL
    from .graphql_client import GQLClient

    client = GQLClient(id_token=id_token, user_email=email)
    user = GQL.get_user_by_email(client, email=email)
    return user.pk


@dataclass
class VerifyOrRefreshIdTokenResult:
    refreshed: bool
    id_token: str
    refresh_token: str
    email: str


def verify_or_refresh_id_token(
    id_token: str, refresh_token: str, force_refresh=False
) -> Optional[VerifyOrRefreshIdTokenResult]:
    """
    Get user object from the id token. Refresh the id token if necessary.
    """
    if force_refresh:
        result = refresh_id_token(refresh_token)
        if user := verify_id_token(result.id_token):
            return VerifyOrRefreshIdTokenResult(
                email=user.email,
                id_token=result.id_token,
                refresh_token=result.refresh_token,
                refreshed=True,
            )
        else:
            return None
    elif user := verify_id_token(id_token):
        return VerifyOrRefreshIdTokenResult(
            email=user.email,
            id_token=id_token,
            refresh_token=refresh_token,
            refreshed=False,
        )
    else:
        result = refresh_id_token(refresh_token)
        user = verify_id_token(result.id_token)
        assert user is not None, "Unable to refresh id token."
        return VerifyOrRefreshIdTokenResult(
            email=user.email,
            id_token=result.id_token,
            refresh_token=result.refresh_token,
            refreshed=True,
        )


@dataclass
class RefreshIdTokenResult:
    id_token: str
    refresh_token: str


def refresh_id_token(
    refresh_token: str,
) -> RefreshIdTokenResult:
    resp = requests.post(
        f"{config.auth_service_host}/refresh-idtoken",
        json={"refreshToken": refresh_token},
    )
    assert resp.status_code == 200, "Unable to refresh id token. " + resp.text
    data = resp.json()
    return RefreshIdTokenResult(
        id_token=data["idToken"], refresh_token=data["refreshToken"]
    )


@dataclass
class VerifyIdTokenResult:
    email: str


def verify_id_token(id_token: str) -> Optional[VerifyIdTokenResult]:
    """Verify the id token. Return the payload if valid, None otherwise."""
    google_cert = get_google_cert()

    try:
        header = jwt.get_unverified_header(id_token)
    except jwt.exceptions.InvalidTokenError as e:
        return None
    if header.get("alg", "") != "RS256":
        return None
    if header.get("kid", "") not in google_cert:
        return None
    google_cert = google_cert.get(header["kid"], "").encode("ASCII")
    if not google_cert:
        return None

    cert_obj = load_pem_x509_certificate(google_cert)
    public_key = cert_obj.public_key()

    try:
        payload = jwt.decode(
            id_token,
            public_key,  # type: ignore
            algorithms=["RS256"],
            audience="fastchargeapi",
        )
        return VerifyIdTokenResult(email=payload["email"])
    except jwt.exceptions.InvalidTokenError as e:
        return None


# def get_or_refresh_token(id_token: str, refresh_token: str) -> RefreshIdTokenResult:
#     """Verifies the id token. Refresh the id token if necessary."""
#     if verify_id_token(id_token) is None:
#         result = _refresh_id_token(refresh_token)
#         write_to_auth_file(id_token=result.id_token, refresh_token=result.refresh_token)
#         return result
#     else:
#         return RefreshIdTokenResult(id_token=id_token, refresh_token=refresh_token)


# def get_or_refresh_id_token_from_auth_file() -> Optional[str]:
#     """Get the id token from the auth file. If the file doesn't exist, return
#     None. If the id token is invalid, refresh it and return the new id token. If
#     unable to refresh, return None."""
#     if auth := _read_auth_file():
#         if token := get_or_refresh_token(auth.id_token, auth.refresh_token):
#             return token.id_token
#     return None

File: fastchargeapi-cli-v2/src/utils/authFile.ts
import { promises as fs } from "fs";
import { glob } from "glob";
import { decodeProtectedHeader, importX509, errors as joseErrors, jwtVerify } from "jose";
import fetch from "node-fetch";
import os from "os";
import path from "path";
import { graphql } from "src/__generated__/gql";
import { authServiceHost } from "src/env";
import { getGQLClient } from "src/graphqlClient";
import { tiChecker } from "src/tiChecker";
import { AuthFileContent, RefreshIdTokenResult, VerifyIdTokenResult } from "src/types/authFile";

export async function getGoogleCert(): Promise<{ [hash: string]: string }> {
  const response = await fetch(
    "https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com"
  );
  return (await response.json()) as { [hash: string]: string };
}

export function getAuthFilePath(profile?: string): string {
  const home = os.homedir();
  if (profile) {
    return path.join(home, ".fastcharge", `auth.${profile}.json`);
  }
  return path.join(home, ".fastcharge", "auth.json");
}

export async function listAuthFiles(): Promise<string[]> {
  const home = os.homedir();
  return await glob([path.join(home, ".fastcharge", "auth.*.json"), path.join(home, ".fastcharge", "auth.json")]);
}

export async function deleteAuthFile(profile?: string) {
  const authFilePath = getAuthFilePath(profile);
  await fs.access(authFilePath);
  await fs.unlink(authFilePath);
}

export async function readAuthFile(profile?: string): Promise<AuthFileContent> {
  const authFilePath = getAuthFilePath(profile);
  const authFileContent = await fs.readFile(authFilePath, "utf-8");
  const auth = tiChecker.AuthFileContent.from(JSON.parse(authFileContent));
  auth.issuer = auth.issuer || "firebase";
  return auth;
}

/**
 * Partially update the auth file context, or create a new auth file with the
 * specified content.
 */
export async function writeToAuthFile(
  profile: string | undefined,
  content: Partial<AuthFileContent>
): Promise<AuthFileContent> {
  const authFilePath = getAuthFilePath(profile);
  await fs.mkdir(path.dirname(authFilePath), { recursive: true });
  let currentAuthFile;
  try {
    currentAuthFile = await readAuthFile(profile);
  } catch {
    currentAuthFile = null;
  }
  // default issuer to firebase
  const newAuthFile = { issuer: "firebase", ...currentAuthFile, ...content };
  await fs.writeFile(authFilePath, JSON.stringify(newAuthFile, null, 2));
  await fs.chmod(authFilePath, 0o600);
  return tiChecker.AuthFileContent.from(newAuthFile);
}

export type VerifyOrRefreshIdTokenResult = {
  refreshed: boolean;
  idToken: string;
  refreshToken: string;
  email: string;
};

export async function refreshIdToken(refreshToken: string): Promise<RefreshIdTokenResult> {
  const resp = await fetch(`${authServiceHost}/refresh-idtoken`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ refreshToken: refreshToken }),
  });

  if (!resp.ok) {
    throw new Error("Unable to refresh id token. " + (await resp.text()));
  }

  return tiChecker.RefreshIdTokenResult.from(await resp.json());
}

export async function verifyIdToken(idToken: string): Promise<VerifyIdTokenResult> {
  const { kid, alg } = decodeProtectedHeader(idToken);

  if (!kid) {
    throw new Error("No kid found in id token.");
  }

  if (!alg) {
    throw new Error("No alg found in id token.");
  }

  const certs = await getGoogleCert();
  const cert = certs[kid];
  if (!cert) {
    throw new Error("No cert found for kid. Google may have rotated the cert and your id token is expired.");
  }

  const key = await importX509(cert, alg);

  try {
    const { payload } = await jwtVerify(idToken, key);
    return tiChecker.VerifyIdTokenResult.from(payload);
  } catch (error) {
    if (error instanceof joseErrors.JWTExpired) {
      throw new Error("Id token expired.");
    }
    throw new Error("Invalid id token. ");
  }
}

export async function verifyOrRefreshIdToken(args: {
  idToken: string;
  refreshToken: string;
  forceRefresh?: boolean;
}): Promise<VerifyOrRefreshIdTokenResult> {
  const { idToken, refreshToken, forceRefresh } = args;
  if (forceRefresh) {
    const newToken = await refreshIdToken(refreshToken);
    const { email } = await verifyIdToken(newToken.idToken);
    return {
      email,
      idToken: newToken.idToken,
      refreshToken: newToken.refreshToken,
      refreshed: true,
    };
  } else {
    try {
      const { email } = await verifyIdToken(idToken);
      return {
        email,
        idToken,
        refreshToken,
        refreshed: false,
      };
    } catch (e) {
      const newToken = await refreshIdToken(refreshToken);
      const { email } = await verifyIdToken(newToken.idToken);
      return {
        email: email,
        idToken: newToken.idToken,
        refreshToken: newToken.refreshToken,
        refreshed: true,
      };
    }
  }
}

/**
 * Gets the id token from the auth file. If the file doesn't exist, returns
 * None. If the id token is issued by Firebase, verifies the token. If the id
 * token is invalid, refreshes it and returns the new id token. If unable to
 * refresh, return None. If the id token is not issued by FastchargeAPI, skips
 * the verification. The new id token and refresh token are written to the auth
 * file.
 */
export async function readOrRefreshAuthFile(args: {
  profile?: string;
  forceRefresh?: boolean;
}): Promise<AuthFileContent | null> {
  const { profile, forceRefresh } = args;
  const auth = await readAuthFile(profile);
  if (auth) {
    if (auth.issuer === "firebase") {
      const user = await verifyOrRefreshIdToken({ ...auth, forceRefresh });
      if (user && user.refreshed) {
        const resp = await getGQLClient(user).query({
          query: graphql(`
            query GetUserPKByEmail($email: Email!) {
              getUserByEmail(email: $email) {
                pk
              }
            }
          `),
          variables: { email: user.email },
        });
        const userPK = resp.data.getUserByEmail.pk;
        return await writeToAuthFile(profile, {
          idToken: user.idToken,
          refreshToken: user.refreshToken,
          email: user.email,
          userPK,
        });
      } else {
        return auth;
      }
    } else if (auth.issuer === "fastchargeapi") {
      return auth;
    } else {
      throw new Error(`Unknown issuer: ${auth.issuer}`);
    }
  }
  return null;
}
